### PR TITLE
9689 q ref as value cond form component

### PR DESCRIPTION
**Pull request by smoyth**

### CODE REVIEW

File: app/javascript/components/conditions/ConditionSetFormField/ConditionFormField/model.js [hunk 1]
```diff
@@ -56,6 +62,31 @@ class ConditionModel {
...
+    // - Allow only these type pairs:
+    //   - textual type -> textual type
+    //   - numeric type -> numeric type
+    //   - select_multiple -> none (literals only, due to ODK restriction)
+    //   - exact match for all other question types (qtypes must be identical)
+    reaction(
+      () => this.leftQingId,
+      (leftQingId) => {
+        if (leftQingId) {
+          const leftQing = this.refableQings.find((qing) => qing.id == leftQingId);
```
Comment (cooperka) [inline app/javascript/components/conditions/ConditionSetFormField/ConditionFormField/model.js]: Various lint errors, I'll fix up briefly after this review

File: app/javascript/components/conditions/ConditionSetFormField/model.js [hunk 2]
```diff
@@ -71,7 +71,10 @@ class ConditionSetModel {
   mapConditionsToStores(conditions) {
     // Only modify if necessary to prevent a cycle.
     if (conditions.some((condition) => !(condition instanceof ConditionModel))) {
-      return conditions.map((condition) => new ConditionModel(condition));
+      return conditions.map((condition) => {
+        condition.refableQings = this.refableQings;
```
Comment (cooperka) [inline app/javascript/components/conditions/ConditionSetFormField/model.js]: `mapConditionsToStores` is here to make sure all conditions are instances of `ConditionModel` and not just plain objects -- this code won't be invoked if the condition starts out as a `ConditionModel`.

I suggest separating concerns and making `refableQings` an expected param in the `ConditionModel` constructor. Then do e.g.

```js
conditions.map((condition) => new ConditionModel({ ...condition, refableQings: this.refableQings });
```

File: app/javascript/components/conditions/ConditionSetFormField/ConditionFormField/component.test.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+describe('with rightSideType as qing', () => {
+  beforeEach(() => {
+    defaultProps.condition.rightSideType = 'qing';
+  });
+
+  describe('with no applicable rightQingOptions', () => {
+    beforeEach(() => {
+      defaultProps.condition.rightQingId = '2';
+    });
```
Comment (cooperka) [inline app/javascript/components/conditions/ConditionSetFormField/ConditionFormField/component.test.js]: This nested prop construction is really nice

**Comment by cooperka** [comment]
Looks like there are also several circular dependencies now that should be addressed (reported by the linter). Let me know if you want help with that.

File: app/assets/javascripts/utils/select2_option_builder.coffee [hunk 4]
```diff
@@ -1,5 +1,8 @@
-select2 = {
-  getAjaxParams: (url, resultsKey = 'results', textKey = 'text') ->
+# Builds options for select2 controls so that common options can be reused without duplication.
+# It's designed like this, instead of as a wrapper of select2, so it can be used in React as well
+# as Backbone.
+class ELMO.Utils.Select2OptionBuilder
```
[the remark this answers is not in the document]
Comment (smoyth) [reply app/assets/javascripts/utils/select2_option_builder.coffee]: Just for consistency. All the other CoffeeScript files are classes.

File: app/javascript/components/conditions/ConditionSetFormField/ConditionFormField/model.js (hunk 1 above)
Comment (smoyth) [reply app/javascript/components/conditions/ConditionSetFormField/ConditionFormField/model.js]: Thanks.

File: app/javascript/components/conditions/ConditionSetFormField/ConditionFormField/component.test.js (hunk 3 above)
Comment (smoyth) [reply app/javascript/components/conditions/ConditionSetFormField/ConditionFormField/component.test.js]: Thanks rspec!

File: app/javascript/components/conditions/ConditionSetFormField/model.js (hunk 2 above)
Comment (smoyth) [reply app/javascript/components/conditions/ConditionSetFormField/model.js]: Good catch. I went with something similar to what you suggested. Hopefully you like it :)

**Pull request merged**